cupy/_logic/truth.py

```diff
@@ -98,26 +120,99 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
     return v1 == v2 if invert else v1 != v2
 
 
+def intersect1d(arr1, arr2, assume_unique=False, return_indices=False):
+    """Find the intersection of two arrays.
+    Returns the sorted, unique values that are in both of the input arrays.
+
+    Parameters
+    ----------
+    arr1, arr2 : cupy.ndarray
+        Input arrays. Arrays will be flattened if they are not in 1D.
+    assume_unique : bool
+        By default, None. If set True, the input arrays will be
+        assumend to be unique, which speeds up the calculation. If set True,
+        but the arrays are not unique, incorrect results and out-of-bounds
+        indices could result.
+    return_indices : bool
+       By default, False. If True, the indices which correspond to the
+       intersection of the two arrays are returned.
+
+    Returns
+    -------
+    intersect1d : cupy.ndarray
+        Sorted 1D array of common and unique elements.
+    comm1 : cupy.ndarray
+        The indices of the first occurrences of the common values
+        in `arr1`. Only provided if `return_indices` is True.
+    comm2 : cupy.ndarray
+        The indices of the first occurrences of the common values
+        in `arr2`. Only provided if `return_indices` is True.
+
+    See Also
+    --------
+    numpy.intersect1d
+
+    """
+    if not assume_unique:
+        if return_indices:
+            arr1, ind1 = cupy.unique(arr1, return_index=True)
+            arr2, ind2 = cupy.unique(arr2, return_index=True)
+        else:
+            arr1 = cupy.unique(arr1)
+            arr2 = cupy.unique(arr2)
+    else:
+        arr1 = arr1.ravel()
+        arr2 = arr2.ravel()
+
+    aux = cupy.concatenate((arr1, arr2))
+    if return_indices:
+        aux_sort_indices = cupy.argsort(aux)
+        aux = aux[aux_sort_indices]
+    else:
+        aux.sort()
+
+    mask = aux[1:] == aux[:-1]
```

Comment:
Who about using the searchsorted as mentioned in https://github.com/cupy/cupy/pull/4018#discussion_r495790724, 

I had also implemented it in PR #6407.

Reply:
There seems to be not significant difference in performance.

- benchmark:
```py
import cupy
from cupyx.profiler import benchmark


size = (16, 4096, 4096)
scale = 100000000
x1 = cupy.testing.shaped_random(size, scale=scale, dtype='l', seed=0)
x2 = cupy.testing.shaped_random(size, scale=scale, dtype='l', seed=1)

for return_indices in [False, True]:
    perf = benchmark(
        cupy.intersect1d,
        (x1, x2),
        {'return_indices': return_indices},
        n_repeat=100,
    )
    print(perf)
```

- sorting
```
intersect1d         :    CPU:111609.020 us   +/-270.427 (min:111209.120 / max:113111.656) us     GPU-0:113952.656 us   +/-266.501 (min:113556.480 / max:115440.643) us
intersect1d         :    CPU:264716.717 us   +/-128.043 (min:264499.548 / max:265360.571) us     GPU-0:271108.404 us   +/-104.410 (min:270897.156 / max:271538.177) us
```

- searchsorted
```
intersect1d         :    CPU:109051.820 us   +/-196.413 (min:108677.675 / max:109495.282) us     GPU-0:110463.191 us   +/-195.613 (min:110091.263 / max:110893.059) us
intersect1d         :    CPU:244261.040 us   +/-121.246 (min:244080.101 / max:244808.975) us     GPU-0:251270.912 us   +/-112.754 (min:251099.136 / max:251734.009) us
```

Reply:
`searchsorted` seems 3~5% better than `sorting`? Why not use it if it's already implemented? 😄